src/vm/vm_memory/memory.rs

```diff
@@ -129,6 +133,81 @@ impl Memory {
         ))
     }
 
+    /// Relocates the memory according to the relocation rules and clears `self.relocaction_rules`.
+    pub fn relocate_memory(&mut self) -> Result<(), MemoryError> {
+        if self.relocation_rules.is_empty() {
+            return Ok(());
+        }
+
+        let mut prev_data = Vec::new();
+        let mut prev_temp_data = Vec::new();
+        swap(&mut self.data, &mut prev_data);
+        swap(&mut self.temp_data, &mut prev_temp_data);
+
+        let data_iter = prev_data
+            .into_iter()
+            .enumerate()
+            .flat_map(|(segment_index, segment_data)| {
+                segment_data
+                    .into_iter()
+                    .enumerate()
+                    .map(move |(cell_offset, cell_data)| {
+                        (
+                            Relocatable::from((segment_index as isize, cell_offset)),
+                            cell_data,
+                        )
+                    })
+            })
+            .chain(prev_temp_data.into_iter().enumerate().flat_map(
+                |(segment_index, segment_data)| {
+                    let segment_index = -(segment_index as isize) - 1;
+                    segment_data
+                        .into_iter()
+                        .enumerate()
+                        .map(move |(cell_offset, cell_data)| {
+                            (
+                                Relocatable::from((segment_index as isize, cell_offset)),
+                                cell_data,
+                            )
+                        })
+                },
+            ));
+        for (addr, value) in data_iter {
+            let value = match value {
+                Some(x) => x,
+                None => continue,
+            };
+
+            let mut new_addr: Relocatable = self
+                .relocate_value(&MaybeRelocatable::RelocatableValue(addr))?
+                .into_owned()
+                .try_into()?;
+            let new_value = self.relocate_value(&value)?.into_owned();
+
+            let target_memory = match new_addr.segment_index.is_negative() {
+                false => &mut self.data,
+                true => {
+                    new_addr.segment_index = -(new_addr.segment_index + 1);
+                    &mut self.temp_data
+                }
+            };
```

Comment:
```suggestion
            if new_addr.segment_index.is_negative() {
                new_addr.segment_index = -(new_addr.segment_index + 1);
            }
            let target_memory = &mut self.temp_data;
```

Reply:
Doesn't this change the functionality? The idea was that since starknet had unified memory, then we should be able to work with both normal and temporary segments.